### ISSUE TITLE
Nats defered message

### ISSUE THREAD

**Issue by Julienhardis** [nats-io/nats-server#3403]
Hello,

we need to defer the delivery of message like it is described in this discussion :

https://github.com/nats-io/nats-server/discussions/2731

(here too : https://github.com/nats-io/nats-server/issues/2846)

I did some investigations about it in issues or discussions and it seems that it's a current need.

Are you considering this in a future release ?

For your information there is a fork on this subject :

https://github.com/nats-io/nats-server/commit/c0d0615ca3fd4708bfe22d8a7e3d8cafb11dad4f

Thank you by advance.

Julien

**Comment by alifpay** [comment]
there is package asyncjobs, it has what you want and uses nats

https://github.com/choria-io/asyncjobs

**Comment by Julienhardis** [comment]
> 

thank you, i'm going to read the documentation.

**Comment by abelkuruvilla** [comment]
@alifpay I think the asyncjobs misses the key feature requested here, i.e, run send a message after a delay of specified time. The server should send the message to subscriber clients only after the specified time/duration has passed.
@Julienhardis I am also looking for this. Have you had any luck finding any solution ?

**Comment by derekcollison** [comment]
We have had a few requests for this, but not a current priority unless someone steps up and wants to sponsor the work.

**Comment by vposhamod** [comment]
Maybe I did not understand the requirement correctly, doesn't Jetstream already give us NakWithDelay with which consumer can write this logic by reading the payload?

**Comment by johnlcox** [comment]
I'd also liked to see this feature so that a message can be delayed/deferred at publish time rather then need to consume and then nak. We have a use case where we are updating 2 systems -- a JetStream stream and a Cassandra table. We publish to the stream first and the Cassandra table second as the Cassandra table is the source of truth. The consumer of the stream will then reconcile the two to filter out messages that may have failed to get inserted into Cassandra.

Without the delayed message on publish, the consumer is likely to fetch messages before the Cassandra write has actually finished. The consumer will either need an artificial wait or the consumer will need to nak most messages to be redelivered after a time we expect Cassandra to have been updated.

**Comment by MauriceVanVeen** [comment]
This is now possible with scheduled messages since 2.12:
https://nats.io/blog/nats-server-2.12-release/#delayed-message-scheduling
